docs/proposal/20240308-vpc-infra-creation.md

```diff
@@ -0,0 +1,393 @@
+# Dynamically create infrastructure required for VPC cluster
+
+## Motivation
+Currently, inorder to create a VPC cluster using cluster api we need to create
+resources as prerequisites. This involves work which is limiting the true
+capabilities of cluster api.
+
+
+## Goal
+1. Dynamically create the required cloud resources as a part of cluster
+   creation.
+
+
+## Proposal
+
+
+### Cluster API VPC cluster components
+
+![vpc-cluster-components.png](../images/vpc-cluster-components.png)
+
+
+### Proposed API changes
+
+
+#### VPCCluster improvements
+```go
+// IBMVPCClusterSpec defines the desired state of IBMVPCCluster.
+type IBMVPCClusterSpec struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// controlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint capiv1beta1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+	// controlPlaneLoadBalancer is optional configuration for customizing control plane behavior.
+	// +optional
+	ControlPlaneLoadBalancer []*VPCLoadBalancerSpec `json:"controlPlaneLoadBalancer,omitempty"`
+
+	// cosInstance is the IBM COS instance to use for cluster resources.
+	COSInstance *CosInstance `json:"cosInstance,omitempty"`
+
+	// networkSpec represents the VPC network to use for the cluster.
+	NetworkSpec *VPCNetworkSpec `json:"networkSpec,omitempty"`
+
+	// region defines the IBM Cloud Region the cluster resources will be deployed in.
+	Region string `json:"region"`
+
+	// resourceGroup defines the IBM Cloud resource group where the cluster resources should be created.
+	ResourceGroup string `json:"resourceGroup"`
+
+	// The Name of VPC. 
+	// Deprecated: use NetworkSpec instead.
```

Comment:
any specific reason for moving `VPC` name inside `NetworkSpec` ?
cc @Karthik-K-N

Reply:
VPC is considered a part of the Network definition, which in VPC's case, can be defined in a unique ResourceGroup, along with the VPC's Subnets, SecurityGroups, etc. So those related resources seemed to fit better within a common type.

Using a NetworkSpec seems to follow other platform's design, rather than a 'flat' VPCCluster struct.